pageserver/src/tenant/timeline/heatmap_layers_downloader.rs

```diff
@@ -0,0 +1,115 @@
+
+use futures::StreamExt;
+use http_utils::error::ApiError;
+use std::sync::{Arc, Mutex};
+
+use super::Timeline;
+
+// This status is not strictly necessary now, but gives us a nice place
+// to store progress information if we ever wish to expose it.
+pub(super) enum HeatmapLayersDownloadStatus {
+    InProgress,
+    Complete,
+}
+
+pub(super) struct HeatmapLayersDownloader {
+    handle: tokio::task::JoinHandle<()>,
+    status: Arc<Mutex<HeatmapLayersDownloadStatus>>,
+}
+
+impl HeatmapLayersDownloader {
+    fn new(
+        timeline: Arc<Timeline>,
+        concurrency: usize,
+    ) -> Result<HeatmapLayersDownloader, ApiError> {
+        let tl_guard = timeline.gate.enter().map_err(|_| ApiError::Cancelled)?;
+        let status = Arc::new(Mutex::new(HeatmapLayersDownloadStatus::InProgress));
+
+        let handle = tokio::task::spawn({
+            let status = status.clone();
+            async move {
+                let _guard = tl_guard;
+
+                scopeguard::defer! {
+                    *status.lock().unwrap() = HeatmapLayersDownloadStatus::Complete;
+                }
+
+                let Some(heatmap) = timeline.generate_heatmap().await else {
+                    tracing::info!("Heatmap layers download failed to generate heatmap");
+                    return;
+                };
+
+                tracing::info!(
+                    resident_size=%timeline.resident_physical_size(),
+                    heatmap_leayrs=%heatmap.layers.len(),
+                    "Starting heatmap layers download"
+                );
+
+                let stream = futures::stream::iter(heatmap.layers.into_iter().map(
+                    |layer| {
+                        let tl = timeline.clone();
+                        async move {
+                            let res = tl.download_layer(&layer.name).await;
+                            if let Err(err) = res {
+                                tracing::warn!(layer=%layer.name,"Failed to download heatmap layer: {err}")
+                            }
+                        }
+                    }
+                )).buffered(concurrency);
+
+                tokio::select! {
+                    _ = stream.collect::<()>() => {
+                        tracing::info!(
+                            resident_size=%timeline.resident_physical_size(),
+                            "Heatmap layers download completed"
+                        );
+                    },
+                    _ = timeline.cancel.cancelled() => {
+                        tracing::info!("Heatmap layers download cancelled. Timeline shutting down.");
+                    }
+                }
+            }
+        });
+
+        Ok(Self { status, handle })
+    }
+
+    fn is_complete(&self) -> bool {
+        matches!(
+            *self.status.lock().unwrap(),
+            HeatmapLayersDownloadStatus::Complete
+        )
+    }
+
+    fn abort(&self) {
+        self.handle.abort();
```

Comment:
Cancellation safety: we're assuming `download_layer` is cancel-safe here, right?  (i.e. it won't leave some rogue I/O in flight writing to the timeline dir).

From a quick look, it seems like that might be the case from how the layer download works internally (there's at least one "spawn" in there), but would be good to be sure + have a comment to that effect.

Reply:
It's not, but there's not much that we can do without reworking the layer download mechanism.
The layer download itself is done in a separate task and that doesn't support cancellation.

Reading through the code, this is all fine since we don't touch any internal state after spawning the task.
It's brittle though. I'll add a gate to make sure that all outstanding downloads complete on abort.

Reply:
Some refinements in https://github.com/neondatabase/neon/pull/10788/commits/9f1648fee403e1623a1504fef4789bcf0af3a68a